### ISSUE TITLE
directory from which tsconfig.json was read 

### ISSUE THREAD

**Issue by craigphicks** [nonara/ts-patch#49]
Let's call it configFileDir only for the scope of this post.  

In `ttypescript`, the path of the tsconfig.json filepath was attached to compilerOptions and could be read from a transform with 
```
program.getCompilerOptions().configFile
```
so
```
path.dirname(program.getCompilerOptions().configFile)
```
gave the value of configFileDir.  
However that added feature (added relative to native typescript) isn't in `ts-patch`.

One level up in the `ts-patch` function `createProgram` there is a local variable `projectDir`, with this code:
```
// TODO - Replace all process.cwd with attempting to first resolve from tsConfigFile dir & bump major version
var projectDir = process.cwd();
```
but since that fix isn't in, and because `projectDir` passed by one means or another,  it isn't possible to run `tsc` from outside configFileDir using 
```
tsc -p <configFileDir>/tsconfig.json
```
That limitation isn't mentioned in the documentation.

I assume it's fixed in v2, so looking forward to the beta version!! :+1:

**Comment by craigphicks** [comment]
Clearly there were some misunderstandings in my post, and thanks for your helpful pointers.
Indeed with `ts-expose-internals` it(*) can be seen, and that's what I had using in the `ttypescript` case.
(* `program.getCompilerOptions().configFilePath`)

However, that seems to be a very user-unfriendly way for a transform writer to have to get *configFileDir*, a.k.a. *projectDir*.
In addition, internals are by definition unstable - although this particular item is almost certainly not going to change,

Wouldn't it be user-friendlier just to pass *projectDir* as a member of `ProgramTransformerExtras`? 

I think opening a new feature suggestion issue would be too much overlap.  It would be better to relabel this topic as a feature suggestion.

**Comment by nonara** [comment]
So it turns out that I had fixed the issue. I just forgot to remove the todo. `tsc -p` should work fine.

> that seems to be a very user-unfriendly way for a transform writer to have to get configFileDir, a.k.a. projectDir.

Welcome to the world of working with the TS compiler ☺ In some time, you'll get comfortable working with the internals. I tried to avoid them for awhile, as everyone does starting out, but you quickly realize that this isn't feasible. Interestingly, it's also not necessarily any more or less 'safe', in terms of things breaking.

Some things to consider:

1. The TS team, regarding semantic versioning, says that they do not use it because in compilers, *every change can be a breaking change*
2. Some internal things are rock solid and very likely not getting changed (like configFilePath), where some public things do change (like transformer factory)
3. I recently had a function we rely on for a very widely used library get changed. (It was not marked internal!) I don't believe the change was even documented in release notes, because it wasn't widely known about.

This is simply the reality of working with these things. For this reason, it's our prerogative as developers to make careful decisions on which internal or 'public' methods and properties to use, opting to try to use those which are least likely to change. A good way to do that is to see how integral (or interwoven) it is into other areas of the compiler.

> Wouldn't it be user-friendlier just to pass projectDir as a member of ProgramTransformerExtras?

`projectDir` is really just a local variable whose sole purpose is to properly resolve the transformer paths listed in `tsconfig.json -> plugins`. The TODO comment above it should be removed.

Also noteworthy is that projectDir gains its information by using `configFilePath`. This is simply the best route for doing it. I did include a fallback method, however, but it would only ever need to be applied if no config file was present, which cannot happen with `tsc` if there are plugins present.

You can see that code here: 

https://github.com/nonara/ts-patch/blob/9b130bc8fd1a5bf3e7b33f796990ed7383aa5449/src/patch/lib/createProgram.ts#L31-L47

**That being said**, I do see value in making an easily accessible property for `projectDir`, so I've added a note to expose it, the config file, and the config file path in `v2`'s context.

> Disable process.chdir() in transforms by overwriting via process.chdir=0

As I mentioned, cwd is no longer relied upon for `tsc`, so there would be no effective purpose. 

See: https://github.com/nonara/ts-patch/blob/9b130bc8fd1a5bf3e7b33f796990ed7383aa5449/src/patch/lib/createProgram.ts#L108-L115

Also, I don't like the idea of overriding it. If a transformer decided to do this, for some reason, it could likely break other areas of the compiler unrelated to ts-patch.

## TLDR

Use `configFilePath`. That's what those of us working with the compiler use for our major projects. Feel free to borrow my code above if you'd like the fallback option.
